### ISSUE TITLE
net/eth/drivers/eth_mcux: TCP client mode is broken

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1577]
**_Reported by Flavio Santes:_**

After applying https://gerrit.zephyrproject.org/r/#/c/11176/ and trying to run the MQTT Publisher application (tcp client application) the application is unable to connect to the server:

[dev/eth_mcux] [DBG] eth_0_init: MAC 00:04:9f:0b:51:e2
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=0
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=1
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=2
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=3
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
net_context_connect error
Is the server (broker) up and running?
[publisher:240] network_setup: -60 <ERROR>

Bye!
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=6
[dev/eth_mcux] [DBG] eth_mcux_phy_event: phy_state=4 

When the patch is reverted (or before it was applied) the output is:

[dev/eth_mcux] [INF] eth_0_init: Enabled 100M full-duplex mode.
[dev/eth_mcux] [DBG] eth_0_init: MAC 00:04:9f:fb:a9:aa
[publisher:240] network_setup: 0 <OK>
[publisher:265] mqtt_init: 0 <OK>
[connect_cb:87] user_data: CONNECTED
[try_to_connect:219] mqtt_tx_connect: 0 <OK>
[publisher:283] try_to_connect: 0 <OK>
[publisher:292] mqtt_tx_pingreq: 0 <OK>
[publisher:297] mqtt_tx_publish: 0 <OK>
[publish_cb:155] <MQTT_PUBACK> packet id: 15414, user_data: PUBLISH
[publisher:302] mqtt_tx_publish: 0 <OK>
[publish_cb:155] <MQTT_PUBREC> packet id: 54656, user_data: PUBLISH
[publish_cb:155] <MQTT_PUBCOMP> packet id: 54656, user_data: PUBLISH

(Imported from Jira ZEP-1724)

**Comment by nashif** [comment]
**_by Anas Nashif:_**

Probably duplicate of ZEP-1715

**Comment by nashif** [comment]
**_by Marcus Shawcroft:_**

The trace shows the eth_mcux driver starting.  The PHY driver starts up but link speed and duplex autonegotiation has not completed. At this point the application attempts to create a tcp connection which fails in net_context_connect() because there is no network yet.  Either
1) The ethernet cable was not connected and hence there was no network.
2) (more likely) the application is racing against the ethernet driver, phy setup is slow, the application is quick off the line.

This looks different to GH-1568, (because the trace in 1715 suggests the network link is up at full duplex 100M before the connection reported as failed.)


I don't believe this is an issue with the network driver.  The recent change in the network driver to mange link up/down simply exposes a latent issue in the stack where by applications simply assume a network is up and running and do not deal with the possibility that a network is not yet started or that the network is down (either transiently or permanently).

I suggest that the application itself should
1) Detect a dropped TCP connection and have a mechanism to re-establish that connection
2) Wait for the network to come up before attempting initial TCP  setup.  This can be acheived by subscribing to events from the network management layer.

_Jukka Rissanen_  and _Tomasz Bursztyka_  feel free to redirect if you think these suggestions are inappropriate.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

It just occurred to me - what if GH-1532?focusedCommentId=16700&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-16700 is another manifestation of the same underlying issue. But the ping issue from it, I saw before recent PHY changes. The UDP delay issue - don't remember seeing.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

Confirmed my hypothesis from the previous comment. With such condition, when UDP lags behind, TCP doesn't work at all. Note that echo server reports that it received my TCP data and that it sent reply.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

guys, what is the status here? It is not clear if this is still an issue or not

**Comment by nashif** [comment]
**_by Flavio Santes:_**

Perhaps Marcus or Paul may give us an update.

**Comment by nashif** [comment]
**_by Paul Sokolovsky:_**

I didn't yet have chance to triage a TCP client issue.

**Comment by nashif** [comment]
**_by jiequan wu:_**

I tested the latest commit on master(commit 27e4fa0d7f) and rc2(commit 46a38ee4), this issue is still exist.

**Comment by nashif** [comment]
**_by Leandro Pereira:_**

I've submitted a patch on Gerrit, but for some reason I couldn't actually test it on a FRDM board.  What this patch does is add a net_event_wait() function that will block the thread for some time until an event happens (which could be an interface coming up or down).

The change in the MQTT Publisher sample is here: https://gerrit.zephyrproject.org/r/#/c/12165/
And the change in the net subsys is here: https://gerrit.zephyrproject.org/r/#/c/12164

_Marcus Shawcroft_ , is this the kind of thing you had in mind in your comment?

**Comment by nashif** [comment]
**_by Anas Nashif:_**

_Leandro Pereira_  If I am not mistaken this is already being done in the net branch by Marcus. I actually was able to overrcome this issue by just sleeping for a few seconds to let the device come up. This will need to be done cleanly in the 1.8 timeframe.

**Comment by nashif** [comment]
**_by Anas Nashif:_**

TCP client is not broken at all, it works just fine, we need another bug or story to track device management and pending on device status. It is wrong to assume the network is up and ready for connections.

**Comment by nashif** [comment]
**_by jiequan wu:_**

On v1.7.0(commit id: bdc3107d5c47d9c), the mqtt_publish sample app works well on FRDM K64F.

**Comment by nashif** [comment]
Blocks GH-734

**Comment by nashif** [comment]
Duplicate of GH-1568
